src/frontend/src/lib/components/signer/SignerConsentMessage.svelte

```diff
@@ -0,0 +1,153 @@
+	import { i18n } from '$lib/stores/i18n.store';
+	import { SIGNER_CONTEXT_KEY, type SignerContext } from '$lib/stores/signer.store';
+	import { toastsError } from '$lib/stores/toasts.store';
+
+	const {
+		consentMessagePrompt: { payload, reset: resetPrompt },
+		callCanisterPrompt: { reset: resetCallCanisterPrompt }
+	} = getContext<SignerContext>(SIGNER_CONTEXT_KEY);
+
+	let approve: ConsentMessageApproval | undefined;
+	$: approve = $payload?.status === 'result' ? $payload?.approve : undefined;
+
+	let reject: Rejection | undefined;
+	$: reject = $payload?.status === 'result' ? $payload?.reject : undefined;
+
+	let consentInfo: icrc21_consent_info | undefined;
+	$: consentInfo = $payload?.status === 'result' ? $payload?.consentInfo : undefined;
+
+	let loading = false;
+	let displayMessage: string | undefined;
+
+	const onPayload = () => {
+		if ($payload?.status === 'loading') {
+			displayMessage = undefined;
+			loading = true;
+
+			// In case the relying party has not closed the window between the last call and requesting a new call, we reset the call canister prompt; otherwise, we might display both the previous result screen and the consent message.
+			// Note that the library handles the case where the relying party tries to submit another call while a call is still being processed. Therefore, we cannot close a prompt here that is not yet finished.
+			resetCallCanisterPrompt();
+
+			return;
+		}
+
+		loading = false;
+
+		if ($payload?.status === 'error') {
+			toastsError({
+				msg: { text: $i18n.signer.consent_message.error.retrieve },
+				err: $payload.details
+			});
+			return;
+		}
+
+		displayMessage =
+			nonNullish(consentInfo) && 'GenericDisplayMessage' in consentInfo.consent_message
+				? consentInfo.consent_message.GenericDisplayMessage
+				: undefined;
+	};
+
+	$: $payload, onPayload();
+
+	type Text = { title: string; content: string } | undefined;
+
+	// We try to split the content and title because we received a chunk of unstructured text from the canister. This works well for the ICP ledger, but we will likely need to iterate on it. There are a few tasks documented in the backlog.
+	const mapText = (markdown: string | undefined): Text => {
+		if (isNullish(markdown)) {
+			return undefined;
+		}
+
+		const [title, ...rest] = markdown.split('\n');
+
+		return {
+			title: title.replace(/^#+\s*/, '').trim(),
+			content: (rest ?? []).join('\n')
+		};
+	};
+
+	let text: Text;
+	$: text = mapText(displayMessage);
+
+	const onApprove = () => {
+		if (isNullish(approve)) {
+			toastsError({
+				msg: { text: $i18n.signer.consent_message.error.no_approve_callback }
+			});
+
+			resetPrompt();
+			return;
+		}
+
+		approve?.();
+		resetPrompt();
+	};
+
+	const onReject = () => {
+		if (isNullish(reject)) {
+			toastsError({
+				msg: { text: $i18n.signer.consent_message.error.no_reject_callback }
+			});
+
+			resetPrompt();
+			return;
+		}
+
+		reject?.();
```

Comment:
the coalescing `?` is not needed here, right?

Reply:
Absolutely right. Thanks.